skpro/distributions/base/_base_array.py

```diff
@@ -0,0 +1,1817 @@
+        or ``my_distribution.iat[row_index, col_index]``
+        subset ``my_distribution`` to the row selected
+        by ``row_index``, col by ``col_index``,
+        to exactly the same col/rows as ``pandas`` ``iat`` would subset
+        rows in ``my_distribution.index`` and columns in ``my_distribution.columns``.
+        """
+        return _Indexer(ref=self, method="_iat")
+
+    @property
+    def at(self):
+        """Integer location indexer, for single index.
+
+        Use ``my_distribution.at[index]`` for ``pandas``-like row/column subsetting
+        of ``BaseDistribution`` descendants.
+
+        ``index`` can be any ``pandas`` ``at`` compatible index subsetter.
+
+        ``my_distribution.at[index]``
+        or ``my_distribution.at[row_index, col_index]``
+        subset ``my_distribution`` to the row selected
+        by ``row_index``, col by ``col_index``,
+        to exactly the same col/rows as ``pandas`` ``at`` would subset
+        rows in ``my_distribution.index`` and columns in ``my_distribution.columns``.
+        """
+        return _Indexer(ref=self, method="_at")
+
+    @property
+    def shape(self):
+        """Shape of self, a pair (2-tuple)."""
+        return self._shape
+
+    @property
+    def ndim(self):
+        """Number of dimensions of self. 2 if array, 0 if scalar."""
+        return len(self._shape)
+
+    def __len__(self):
+        """Length of self, number of rows."""
+        shape = self._shape
+        if len(shape) == 0:
+            return 1
+        return shape[0]
+
+    def head(self, n=5):
+        """Return the first n rows.
+
+        If there are less than n rows in ``self``, returns clone of ``self``.
+
+        For negative n, returns all rows except the last n.
+
+        Parameters
+        ----------
+        n : int, default=5
+            Number of rows to return.
+
+        Returns
+        -------
+        ``self`` subset to the first n rows, i.e., ``self.iloc[0:min(n, len(self))]``
+        """
+        if self.ndim < 2:
+            return self
+        assert isinstance(n, int)
+        N = len(self)
+        if n < 0:
+            n = N - n
+        n = min(n, N)
+        return self.iloc[range(n)]
+
+    def tail(self, n=5):
+        """Return the last n rows.
+
+        If there are less than n rows in ``self``, returns clone of ``self``.
+
+        For negative n, returns all rows except the first n.
+
+        Parameters
+        ----------
+        n : int, default=5
+            Number of rows to return.
+
+        Returns
+        -------
+        ``self`` subset to the last n rows, i.e., ``self.iloc[max(len(self) - n, 0):]``
+        """
+        if self.ndim < 2:
+            return self
+        assert isinstance(n, int)
+        N = len(self)
+        if n < 0:
+            start = n
+        else:
+            start = N - n
+        start = max(0, start)
+        return self.iloc[range(start, N)]
+
+    def _loc(self, rowidx=None, colidx=None):
```

Comment:
I made some changes in this by adding some `pd.Index` to make it work as some of the places were just list.